File: test/js/filter-tree.test.ts
/**
 * @jest-environment jsdom
 */

import {
  createEmptyTree,
  createEmptyGroup,
  createCondition,
  addCondition,
  removeCondition,
  updateCondition,
  addGroup,
  removeGroup,
  updateGroupOperator,
  findNode,
  getGroupDepth,
  countConditions,
  canAddCondition,
  canAddGroup,
  serializeTree,
  parseTree
} from '../assets/js/lib/filter-tree'

describe('Filter Tree Utilities', () => {
  describe('createEmptyTree', () => {
    it('creates a tree with version 1 and empty root group', () => {
      const tree = createEmptyTree()

      expect(tree.version).toBe(1)
      expect(tree.root).toBeDefined()
      expect(tree.root.type).toBe('group')
      expect(tree.root.operator).toBe('and')
      expect(tree.root.children).toEqual([])
      expect(tree.root.id).toBeDefined()
    })
  })

  describe('createEmptyGroup', () => {
    it('creates an empty group with default AND operator', () => {
      const group = createEmptyGroup()

      expect(group.type).toBe('group')
      expect(group.operator).toBe('and')
      expect(group.children).toEqual([])
      expect(group.id).toBeDefined()
    })

    it('creates an empty group with specified OR operator', () => {
      const group = createEmptyGroup('or')

      expect(group.operator).toBe('or')
    })
  })

  describe('createCondition', () => {
    it('creates a condition with defaults', () => {
      const condition = createCondition()

      expect(condition.type).toBe('condition')
      expect(condition.attribute).toBe('')
      expect(condition.operator).toBe('is')
      expect(condition.value).toBe('')
      expect(condition.negated).toBe(false)
      expect(condition.id).toBeDefined()
    })

    it('creates a condition with specified values', () => {
      const condition = createCondition('visit:country', 'is', 'US')

      expect(condition.attribute).toBe('visit:country')
      expect(condition.operator).toBe('is')
      expect(condition.value).toBe('US')
    })
  })

  describe('addCondition', () => {
    it('adds a condition to the root group', () => {
      const tree = createEmptyTree()
      const condition = createCondition('visit:country', 'is', 'US')

      const newTree = addCondition(tree, tree.root.id, condition)

      expect(newTree.root.children).toHaveLength(1)
      expect(newTree.root.children[0]).toEqual(condition)
    })
  })

  describe('removeCondition', () => {
    it('removes a condition from the tree', () => {
      const tree = createEmptyTree()
      const condition = createCondition('visit:country', 'is', 'US')
      const treeWithCondition = addCondition(tree, tree.root.id, condition)

      const newTree = removeCondition(treeWithCondition, condition.id)

      expect(newTree.root.children).toHaveLength(0)
    })
  })

  describe('updateCondition', () => {
    it('updates condition values', () => {
      const tree = createEmptyTree()
      const condition = createCondition('visit:country', 'is', 'US')
      const treeWithCondition = addCondition(tree, tree.root.id, condition)

      const newTree = updateCondition(treeWithCondition, condition.id, { value: 'GB' })
      const updatedCondition = newTree.root.children[0]

      expect(updatedCondition.value).toBe('GB')
      expect(updatedCondition.attribute).toBe('visit:country')
    })
  })

  describe('addGroup', () => {
    it('adds a nested group to the root', () => {
      const tree = createEmptyTree()
      const nestedGroup = createEmptyGroup('or')

      const newTree = addGroup(tree, tree.root.id, nestedGroup)

      expect(newTree.root.children).toHaveLength(1)
      expect(newTree.root.children[0]).toEqual(nestedGroup)
    })
  })

  describe('countConditions', () => {
    it('counts conditions in a flat tree', () => {
      const tree = createEmptyTree()
      const condition1 = createCondition('visit:country', 'is', 'US')
      const condition2 = createCondition('visit:device', 'is', 'Mobile')
      let treeWithConditions = addCondition(tree, tree.root.id, condition1)
      treeWithConditions = addCondition(treeWithConditions, tree.root.id, condition2)

      expect(countConditions(treeWithConditions)).toBe(2)
    })

    it('counts conditions in nested groups', () => {
      const tree = createEmptyTree()
      const condition1 = createCondition('visit:country', 'is', 'US')
      const nestedGroup = createEmptyGroup('and')
      let treeWithGroup = addGroup(tree, tree.root.id, nestedGroup)
      treeWithGroup = addCondition(treeWithGroup, nestedGroup.id, condition1)

      expect(countConditions(treeWithGroup)).toBe(1)
    })
  })

  describe('canAddCondition', () => {
    it('allows adding conditions under the limit', () => {
      const tree = createEmptyTree()

      expect(canAddCondition(tree)).toBe(true)
    })
  })

  describe('canAddGroup', () => {
    it('allows adding groups at root level', () => {
      const tree = createEmptyTree()

      expect(canAddGroup(tree, tree.root.id)).toBe(true)
    })

    it('disallows adding groups beyond max depth', () => {
      // Create a tree that's already at max depth
      let tree = createEmptyTree()

      // Add nested groups to reach depth limit
      let currentGroupId = tree.root.id
      for (let i = 0; i < 4; i++) {
        const nestedGroup = createEmptyGroup('and')
        tree = addGroup(tree, currentGroupId, nestedGroup)
        currentGroupId = nestedGroup.id
      }

      // At depth 4 (0-indexed), we should still be able to add
      expect(canAddGroup(tree, currentGroupId)).toBe(true)

      // Add one more level
      const oneMoreGroup = createEmptyGroup('and')
      tree = addGroup(tree, currentGroupId, oneMoreGroup)

      // Now at depth 5, should be disallowed
      expect(canAddGroup(tree, oneMoreGroup.id)).toBe(false)
    })
  })

  describe('serializeTree and parseTree', () => {
    it('serializes and parses a tree correctly', () => {
      const tree = createEmptyTree()
      const condition = createCondition('visit:country', 'is', 'US')
      const treeWithCondition = addCondition(tree, tree.root.id, condition)

      const serialized = serializeTree(treeWithCondition)
      const parsed = parseTree(serialized)

      expect(parsed).not.toBeNull()
      expect(parsed?.version).toBe(1)
      expect(parsed?.root.children).toHaveLength(1)
      expect(parsed?.root.children[0].attribute).toBe('visit:country')
    })

    it('returns null for invalid JSON', () => {
      const result = parseTree('invalid json')

      expect(result).toBeNull()
    })
  })

  describe('findNode', () => {
    it('finds a condition by ID', () => {
      const tree = createEmptyTree()
      const condition = createCondition('visit:country', 'is', 'US')
      const treeWithCondition = addCondition(tree, tree.root.id, condition)

      const found = findNode(treeWithCondition, condition.id)

      expect(found).toEqual(condition)
    })

    it('returns null for non-existent ID', () => {
      const tree = createEmptyTree()

      const found = findNode(tree, 'non-existent-id')

      expect(found).toBeNull()
    })
  })

  describe('getGroupDepth', () => {
    it('returns 0 for root group', () => {
      const tree = createEmptyTree()

      expect(getGroupDepth(tree, tree.root.id)).toBe(0)
    })

    it('returns correct depth for nested groups', () => {
      let tree = createEmptyTree()
      const nestedGroup = createEmptyGroup('and')
      tree = addGroup(tree, tree.root.id, nestedGroup)

      expect(getGroupDepth(tree, nestedGroup.id)).toBe(1)
    })
  })
})

File: test/plausible/segments/filter_tree_test.exs
defmodule Plausible.Segments.FilterTreeTest do
  use Plausible.DataCase, async: true
  alias Plausible.Segments.FilterTree

  describe "new/0" do
    test "creates an empty filter tree" do
      tree = FilterTree.new()

      assert tree.root_group.operator == :and
      assert tree.root_group.children == []
      assert tree.version == 1
    end
  end

  describe "add_condition/2" do
    test "adds a condition to root group" do
      tree = FilterTree.new()
      condition = %{dimension: "country", operator: :is, values: ["US"]}

      new_tree = FilterTree.add_condition(tree, condition)

      assert length(new_tree.root_group.children) == 1
      child = hd(new_tree.root_group.children)
      assert child.dimension == "country"
      assert child.operator == :is
      assert child.values == ["US"]
    end

    test "adds multiple conditions to root group" do
      tree = FilterTree.new()
      condition1 = %{dimension: "country", operator: :is, values: ["US"]}
      condition2 = %{dimension: "device", operator: :is, values: ["mobile"]}

      tree = FilterTree.add_condition(tree, condition1)
      tree = FilterTree.add_condition(tree, condition2)

      assert length(tree.root_group.children) == 2
    end
  end

  describe "add_group/2" do
    test "adds a nested group to root" do
      tree = FilterTree.new()

      new_tree = FilterTree.add_group(tree, :or)

      assert length(new_tree.root_group.children) == 1
      child = hd(new_tree.root_group.children)
      assert child.operator == :or
      assert child.children == []
    end

    test "adds group to parent group" do
      tree = FilterTree.new()
      tree = FilterTree.add_group(tree, :or)

      [group | _] = tree.root_group.children

      new_tree = FilterTree.add_group(tree, :and, group.id)

      [nested_group | _] = hd(new_tree.root_group.children).children
      assert nested_group.operator == :and
    end
  end

  describe "remove_item/2" do
    test "removes a condition by id" do
      tree = FilterTree.new()
      tree = FilterTree.add_condition(tree, %{dimension: "country", operator: :is, values: ["US"]})
      tree = FilterTree.add_condition(tree, %{dimension: "device", operator: :is, values: ["mobile"]})

      [to_remove | _] = tree.root_group.children
      new_tree = FilterTree.remove_item(tree, to_remove.id)

      assert length(new_tree.root_group.children) == 1
    end
  end

  describe "delete_group/2" do
    test "deletes a nested group" do
      tree = FilterTree.new()
      tree = FilterTree.add_group(tree, :or)
      tree = FilterTree.add_condition(tree, %{dimension: "country", values: ["US"]})

      [group | _] = tree.root_group.children
      new_tree = FilterTree.delete_group(tree, group.id)

      # The condition should remain
      assert length(new_tree.root_group.children) == 1
    end

    test "does not delete root group" do
      tree = FilterTree.new()
      new_tree = FilterTree.delete_group(tree, tree.root_group.id)

      assert new_tree.root_group.id == tree.root_group.id
    end
  end

  describe "update_condition/3" do
    test "updates condition values" do
      tree = FilterTree.new()
      tree = FilterTree.add_condition(tree, %{dimension: "country", operator: :is, values: ["US"]})

      [condition | _] = tree.root_group.children
      new_tree = FilterTree.update_condition(tree, condition.id, %{values: ["UK"]})

      updated = hd(new_tree.root_group.children)
      assert updated.values == ["UK"]
    end

    test "updates condition operator" do
      tree = FilterTree.new()
      tree = FilterTree.add_condition(tree, %{dimension: "country", operator: :is, values: ["US"]})

      [condition | _] = tree.root_group.children
      new_tree = FilterTree.update_condition(tree, condition.id, %{operator: :contains})

      updated = hd(new_tree.root_group.children)
      assert updated.operator == :contains
    end
  end

  describe "change_group_operator/3" do
    test "changes group operator from and to or" do
      tree = FilterTree.new()
      tree = FilterTree.add_group(tree, :and)
      tree = FilterTree.add_condition(tree, %{dimension: "country", values: ["US"]})

      [group | _] = tree.root_group.children
      new_tree = FilterTree.change_group_operator(tree, group.id, :or)

      updated = hd(new_tree.root_group.children)
      assert updated.operator == :or
    end
  end

  describe "get_depth/1" do
    test "returns 1 for flat group" do
      tree = FilterTree.new()
      tree = FilterTree.add_condition(tree, %{dimension: "country", values: ["US"]})

      assert FilterTree.get_depth(tree.root_group) == 1
    end

    test "returns 2 for single nesting" do
      tree = FilterTree.new()
      tree = FilterTree.add_group(tree, :or)

      [group | _] = tree.root_group.children
      tree = FilterTree.add_group(tree, :and, group.id)

      assert FilterTree.get_depth(tree.root_group) == 2
    end
  end

  describe "validate/1" do
    test "returns error for empty tree" do
      tree = FilterTree.new()
      result = FilterTree.validate(tree)

      assert result.valid == false
      assert "Filter tree must have at least one condition" in result.errors
    end

    test "returns valid for tree with conditions" do
      tree = FilterTree.new()
      tree = FilterTree.add_condition(tree, %{dimension: "country", operator: :is, values: ["US"]})

      result = FilterTree.validate(tree)

      assert result.valid == true
      assert result.errors == []
    end

    test "returns error for missing dimension" do
      tree = FilterTree.new()
      tree = FilterTree.add_condition(tree, %{dimension: "", operator: :is, values: ["US"]})

      result = FilterTree.validate(tree)

      assert result.valid == false
      assert Enum.any?(result.errors, &String.contains?(&1, "missing dimension"))
    end

    test "returns error for missing value" do
      tree = FilterTree.new()
      tree = FilterTree.add_condition(tree, %{dimension: "country", operator: :is, values: []})

      result = FilterTree.validate(tree)

      assert result.valid == false
      assert Enum.any?(result.errors, &String.contains?(&1, "missing value"))
    end
  end

  describe "serialize/1" do
    test "serializes simple filter" do
      tree = FilterTree.new()
      tree = FilterTree.add_condition(tree, %{dimension: "country", operator: :is, values: ["US"]})

      serialized = FilterTree.serialize(tree)

      assert serialized == [[:is, "country", ["US"]]]
    end

    test "serializes multiple filters as AND" do
      tree = FilterTree.new()
      tree = FilterTree.add_condition(tree, %{dimension: "country", values: ["US"]})
      tree = FilterTree.add_condition(tree, %{dimension: "device", values: ["mobile"]})

      serialized = FilterTree.serialize(tree)

      assert serialized == [
        [:is, "country", ["US"]],
        [:is, "device", ["mobile"]]
      ]
    end

    test "serializes nested groups" do
      tree = FilterTree.new()
      tree = FilterTree.add_condition(tree, %{dimension: "country", values: ["US"]})
      tree = FilterTree.add_group(tree, :or)
      [group | _] = tree.root_group.children
      tree = FilterTree.add_condition(tree, %{dimension: "device", values: ["mobile"]}, group.id)

      serialized = FilterTree.serialize(tree)

      assert serialized == [
        [:is, "country", ["US"]],
        [:or, [[:is, "device", ["mobile"]]]]
      ]
    end
  end

  describe "deserialize/1" do
    test "deserializes simple filter" do
      filters = [[:is, "country", ["US"]]]

      tree = FilterTree.deserialize(filters)

      assert length(tree.root_group.children) == 1
      condition = hd(tree.root_group.children)
      assert condition.dimension == "country"
      assert condition.values == ["US"]
    end

    test "deserializes empty array to empty tree" do
      tree = FilterTree.deserialize([])

      assert tree.root_group.children == []
    end

    test "deserializes nested groups" do
      filters = [
        [:is, "country", ["US"]],
        [:or, [
          [:is, "device", ["mobile"]],
          [:is, "browser", ["Chrome"]]
        ]]
      ]

      tree = FilterTree.deserialize(filters)

      assert length(tree.root_group.children) == 2
    end
  end
end
